internal/spec/spec.go

```diff
@@ -0,0 +1,143 @@
+			return nil, fmt.Errorf("failed to unmarshal extra specs: %w", err)
+		}
+	}
+
+	return spec, nil
+}
+
+type extraSpecs struct {
+	Ocpus          float32  `json:"ocpus"`
+	MemoryInGBs    float32  `json:"memory_in_gbs"`
+	BootVolumeSize int64    `json:"boot_volume_size"`
+	SSHPublicKeys  []string `json:"ssh_public_keys"`
+}
+
+func GetRunnerSpecFromBootstrapParams(cfg *config.Config, data params.BootstrapInstance, controllerID string) (*RunnerSpec, error) {
+	tools, err := util.GetTools(data.OSType, data.OSArch, data.Tools)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tools: %s", err)
+	}
+
+	extraSpecs, err := newExtraSpecsFromBootstrapData(data)
+	if err != nil {
+		return nil, fmt.Errorf("error loading extra specs: %w", err)
+	}
+
+	spec := &RunnerSpec{
+		AvailabilityDomain: cfg.AvailabilityDomain,
+		CompartmentID:      cfg.CompartmentId,
+		SubnetID:           cfg.SubnetID,
+		NsgID:              cfg.NsgID,
+		ControllerID:       controllerID,
+		Tools:              tools,
+		BootstrapParams:    data,
+	}
+
+	spec.MergeExtraSpecs(extraSpecs)
+	spec.SetUserData()
+
+	return spec, nil
+}
+
+type RunnerSpec struct {
+	AvailabilityDomain string
+	CompartmentID      string
+	SubnetID           string
+	NsgID              string
+	BootVolumeSize     int64
+	UserData           string
+	ControllerID       string
+	Ocpus              float32
+	MemoryInGBs        float32
+	SSHPublicKeys      []string
+	Tools              params.RunnerApplicationDownload
+	BootstrapParams    params.BootstrapInstance
+}
+
+func (r *RunnerSpec) MergeExtraSpecs(extraSpecs *extraSpecs) {
+	if extraSpecs.Ocpus > 0 {
+		r.Ocpus = extraSpecs.Ocpus
+	} else {
+		r.Ocpus = 1
+	}
+	if extraSpecs.MemoryInGBs > 0 {
+		r.MemoryInGBs = extraSpecs.MemoryInGBs
+	} else {
+		r.MemoryInGBs = 4
+	}
+	if extraSpecs.BootVolumeSize > 0 {
+		r.BootVolumeSize = extraSpecs.BootVolumeSize
+	} else {
+		r.BootVolumeSize = 255
+	}
+	if len(extraSpecs.SSHPublicKeys) > 0 {
+		r.SSHPublicKeys = extraSpecs.SSHPublicKeys
+	}
+}
+
+func (r *RunnerSpec) SetUserData() error {
+	customData, err := r.ComposeUserData()
+	if err != nil {
+		return fmt.Errorf("failed to compose userdata: %w", err)
+	}
+
+	if len(customData) == 0 {
+		return fmt.Errorf("failed to generate custom data")
+	}
+
+	asBase64 := base64.StdEncoding.EncodeToString(customData)
+	r.UserData = asBase64
+	return nil
+}
+
+func (r *RunnerSpec) ComposeUserData() ([]byte, error) {
+	switch r.BootstrapParams.OSType {
+	case params.Linux:
+		udata, err := cloudconfig.GetCloudConfig(r.BootstrapParams, r.Tools, r.BootstrapParams.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate userdata: %w", err)
+		}
+		return []byte(udata), nil
+	case params.Windows:
+		udata, err := cloudconfig.GetCloudConfig(r.BootstrapParams, r.Tools, r.BootstrapParams.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate userdata: %w", err)
+		}
+		wrapped := fmt.Sprintf("<powershell>%s</powershell>", udata)
```

Comment:
Does this work on OCI? I was under the impression they were using cloudbase-init.

Reply:
It works, but it is not needed. It works without as well, updated the code.